File: CSharp/Material/Material.cs
using System;

namespace CSharp
{
    public class Material
    {
        public Material()
        {
            Color = Color.White;
            Ambient = 0.1;
            Diffuse = 0.9;
            Specular = 0.9;
            Shininess = 200.0;
            Reflective = 0.0;
            Transparency = 0.0;
            RefractiveIndex = 1.0;
        }

        public Color Color { get; set; }
        public double Ambient { get; set; }
        public double Diffuse { get; set; }
        public double Specular { get; set; }
        public double Shininess { get; set; }
        public double Reflective { get; set; }
        public double Transparency { get; set; }
        public double RefractiveIndex { get; set; }
        public RTPattern Pattern { get; set; }

        public Color Lighting(RTObject obj, ILight light, Point position, Vector eyev, Vector normalv, double intensity)
        {
            Color color;
            if (Pattern != null)
            {
                color = Pattern.PatternAtObject(obj, position);
            }
            else if (obj.Parent != null)
            {
                color = obj.Parent.Material.Lighting(obj.Parent, light, position, eyev, normalv, intensity);
            }
            else
            {
                color = Color;
            }

            var effectiveColor = color * light.Intensity;
            var ambient = effectiveColor * Ambient;
            var samples = light.SamplePoints();

            Color sum = Color.Black;

            foreach (var sample in samples)
            {
                Color diffuse;
                Color specular;

                var lightV = (new Vector(sample - position)).Normalize;
                var lightDotNormal = lightV.Dot(normalv);

                if ((lightDotNormal < 0) || intensity.IsEqual(0.0))
                {
                    diffuse = Color.Black; // inShadow ? Color.Green : Color.Black;
                    specular = Color.Black; // inShadow ? Color.Green : Color.Black;
                }
                else
                {
                    diffuse = effectiveColor * Diffuse * lightDotNormal;
                    var reflectV = (-lightV).Reflect(normalv);
                    var reflectDotEye = reflectV.Dot(eyev);

                    if (reflectDotEye <= 0)
                    {
                        specular = Color.Black;
                    }
                    else
                    {
                        var factor = Math.Pow(reflectDotEye, Shininess);
                        specular = light.Intensity * Specular * factor;
                    }
                }

                sum += diffuse;
                sum += specular;
            }

            return ambient + ((sum / light.Samples) * intensity);
        }

        public override bool Equals(object obj)
        {
            var other = obj as Material;
            return ((other != null) &&
                    (Color.Equals(other.Color)) &&
                    (Ambient.IsEqual(other.Ambient)) &&
                    (Diffuse.IsEqual(other.Diffuse)) &&
                    (Specular.IsEqual(other.Specular)) &&
                    (Shininess.IsEqual(other.Shininess)) &&
                    (Reflective.IsEqual(other.Reflective)) &&
                    (Transparency.IsEqual(other.Transparency)) &&
                    (RefractiveIndex.IsEqual(other.RefractiveIndex)));
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Ambient, Diffuse, Specular, Shininess, Reflective, Transparency, RefractiveIndex);
        }

        public override string ToString()
        {
            return $"{Color}: ({Ambient},{Diffuse},{Specular},{Shininess}, {Reflective}, {Transparency}, {RefractiveIndex})";
        }
    }
}

File: Java/src/main/java/Material/Material.java
package RTC.Material;

import RTC.Math.*;
import RTC.Lights.*;
import RTC.Patterns.*;
import RTC.Shapes.*;

import java.util.Objects;

public class Material {
    public static int MaterialId = 0;

    private int id;
    private Pattern pattern;
    private Color color;
    private double ambient;
    private double diffuse;
    private double specular;
    private double shininess;

    public Material() {
        id = MaterialId++;

        pattern = null;
        color = Color.White;
        ambient = 0.1;
        diffuse = 0.9;
        specular = 0.9;
        shininess = 200.0;
    }

    public Pattern getPattern() {
        return pattern;
    }

    public void setPattern(Pattern p) {
        pattern = p;
    }

    public Color getColor() {
        return color;
    }

    public void setColor(Color c) {
        color = c;
    }
    
    public double getAmbient() {
        return ambient;
    }

    public void setAmbient(double a) {
        ambient = a;
    }

    public double getDiffuse() {
        return diffuse;
    }

    public void setDiffuse(double d) {
        diffuse = d;
    }

    public double getSpecular() {
        return specular;
    }

    public void setSpecular(double s) {
        specular = s;
    }

    public double getShininess() {
        return shininess;
    }

    public void setShininess(double s) {
        shininess = s;
    }

    public Color lighting(Shape obj, PointLight light, Point point, Vector eyev, Vector normalv, boolean inShadow) {
        Color c;
        if (pattern != null) {
            c = pattern.patternAtShape(obj, point);
        } else {
            c = color;
        }

        Color amb, diff, spec;
        Color effectiveColor = c.mul(light.getIntensity());
        Tuple lightv = (light.getPosition().sub(point)).norm();

        amb = effectiveColor.mul(ambient);

        double lightDotNormal = lightv.dot(normalv);
        if (inShadow || (lightDotNormal < 0.0)) {
            diff = Color.Black;
            spec = Color.Black;
        } else {
            diff = effectiveColor.mul(diffuse).mul(lightDotNormal);
            Tuple reflectv = lightv.neg().reflect(normalv);
            double reflectDotEye = reflectv.dot(eyev);
            if (reflectDotEye <= 0.0) {
                spec = Color.Black;
            } else {
                double factor = Math.pow(reflectDotEye, shininess);
                spec = light.getIntensity().mul(specular).mul(factor);
            }
        }

        return amb.add(diff).add(spec);
    }

    @Override
    public boolean equals(Object o)
    {
        if (o == this) {
            return true;
        }

        if ((o == null) || !(o instanceof Material)) {
            return false;
        }

        Material m = (Material) o;
        return color.equals(m.getColor()) &&
               Helper.IsEqual(ambient, m.getAmbient()) &&
               Helper.IsEqual(diffuse, m.getDiffuse()) &&
               Helper.IsEqual(specular, m.getSpecular()) &&
               Helper.IsEqual(shininess, m.getShininess());
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, ambient, diffuse, specular, shininess);
    }

    @Override
    public String toString() {
        return String.format("Material %s %f %f %f %f Id: %d", color.toString(), ambient, diffuse, specular, shininess, id);
    }    
}
